src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs

```diff
@@ -577,57 +579,67 @@ internal static void CreateSymbolicLink(string path, string pathToTarget, bool i
             // throws if the current link file does not exist
             Interop.CheckIo(Interop.Sys.LStat(linkPath, out _), linkPath, isDirectory);
 
-            string? targetPath = GetLinkTarget(linkPath);
-            if (targetPath == null)
-            {
-                // linkPath exists but is not a link
-                return null;
-            }
-
-            //Ensure all paths are fully qualified, by adding a prefix that is relative to the previous path
-            string? prefix;
-            if (PathInternal.IsPartiallyQualified(targetPath))
-            {
-                prefix = Path.GetDirectoryName(linkPath);
-                targetPath = Path.Join(prefix, targetPath);
-            }
+            ValueStringBuilder sb = new(stackalloc char[Interop.PathStackBufferSize]);
+            sb.Append(linkPath);
 
             int maxVisits = returnFinalTarget ? MaxFollowedLinks : 1;
             int visitCount = 1;
-            while (visitCount < maxVisits)
+            while (visitCount <= maxVisits)
             {
-                string? nextPath = GetLinkTarget(targetPath);
-
-                if (nextPath == null)
+                if (!TryGetLinkTarget(ref sb))
                 {
-                    // targetPath does not exist or is not a link
-                    break;
-                }
+                    if (visitCount == 1)
```

Comment:
If I do the changes above, I need to also change this to:

```suggestion
                    if (visitCount == 0)
```